java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java

```diff
@@ -261,7 +267,50 @@
                 String uri = ((JsonPrimitive) params.getArguments().get(0)).getAsString();
                 Position pos = gson.fromJson(gson.toJson(params.getArguments().get(1)), Position.class);
                 return (CompletableFuture)((TextDocumentServiceImpl)server.getTextDocumentService()).superImplementations(uri, pos);
-                
+            case Server.JAVA_SOURCE_FOR: {
+                CompletableFuture<Object> result = new CompletableFuture<>();
+                try {
+                    String sourceUri = ((JsonPrimitive) params.getArguments().get(0)).getAsString();
+                    if (sourceUri == null || !sourceUri.startsWith(SOURCE_FOR)) {
+                        throw new IllegalArgumentException("Invalid uri: " + sourceUri);
+                    }
+                    sourceUri = URLDecoder.decode(sourceUri.substring(SOURCE_FOR.length()), StandardCharsets.UTF_8.toString());
+                    int qIdx = sourceUri.lastIndexOf('?');
+                    int hIdx = sourceUri.lastIndexOf('#');
+                    if (qIdx < 0 || hIdx < 0 || hIdx <= qIdx) {
+                        throw new IllegalArgumentException("Invalid uri: " + sourceUri);
+                    }
+                    String rootUri = sourceUri.substring(0, qIdx);
+                    FileObject root = URLMapper.findFileObject(URI.create(rootUri).toURL());
+                    if (root == null) {
+                        throw new IllegalStateException("Unable to find root: " + rootUri);
+                    }
+                    ElementHandle typeHandle = ElementHandleAccessor.getInstance().create(ElementKind.valueOf(sourceUri.substring(qIdx + 1, hIdx)), sourceUri.substring(hIdx + 1));
+                    CompletableFuture<ElementOpen.Location> location = ElementOpen.getLocation(ClasspathInfo.create(root), typeHandle, typeHandle.getQualifiedName().replace('.', '/') + ".class");
+                    location.exceptionally(ex -> {
+                        result.completeExceptionally(ex);
+                        return null;
+                    }).thenAccept(loc -> {
+                        if (loc != null) {
+                            ShowDocumentParams sdp = new ShowDocumentParams(Utils.toUri(loc.getFileObject()));
+                            Position position = Utils.createPosition(loc.getFileObject(), loc.getStartOffset());
+                            sdp.setSelection(new Range(position, position));
+                            client.showDocument(sdp).thenAccept(res -> {
+                                if (res.isSuccess()) {
+                                    result.complete(null);
+                                } else {
+                                    result.completeExceptionally(new IllegalStateException("Cannot open source for: " + typeHandle.getQualifiedName()));
```

Comment:
The exception msg  is going to be printed to the client, maybe L10n it.

Reply:
A majority of error messages within the language server implementation is not localized.